KEP-961: Bump Statefulset maxUnavailable to beta (#3997)
kubernetes/enhancements: keps/sig-apps/961-maxunavailable-for-statefulset/README.md

easier for reviewers to cite specific portions, and to minimize diff churn on
updates.

[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
-->

The purpose of this enhancement is to implement maxUnavailable for StatefulSet during RollingUpdate.
When a StatefulSet’s `.spec.updateStrategy.type` is set to `RollingUpdate`, the StatefulSet controller
will delete and recreate each Pod in the StatefulSet. The updating of each Pod currently happens one at a time.
With support for `maxUnavailable`, the updating will proceed `maxUnavailable` number of pods at a time.

## Motivation

<!--
This section is for explicitly listing the motivation, goals, and non-goals of
this KEP.  Describe why the change is important and the benefits to users. The
motivation section can optionally provide links to [experience reports] to
demonstrate the interest in a KEP within the wider Kubernetes community.

[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
-->

Consider the following scenarios:-

1. My containers publish metrics to a time series system. If I am using a Deployment, each rolling

> update creates a new pod name and hence the metrics published by this new pod starts a new time series

Reviewer: We shouldn't be using pod names in metrics, the cardinality is way too high.

...
```

### Test Plan

<!--
**Note:** *Not required until targeted at a release.*
The goal is to ensure that we don't accept enhancements with inadequate testing.

All code is expected to have adequate tests (eventually with coverage
expectations). Please adhere to the [Kubernetes testing guidelines][testing-guidelines]
when drafting this test plan.

[testing-guidelines]: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
-->

[x] I/we understand the owners of the involved components may require updates to
existing tests to make this code solid enough prior to committing the changes necessary
to implement this enhancement.

##### Prerequisite testing updates

<!--
Based on reviewers feedback describe what additional tests need to be added prior
implementing this enhancement to ensure the enhancements have also solid foundations.
-->

> No.

Reviewer 2: This no doesn't mean anything, this section should outline any  additional tests that are needed, since we don't have any, yet, just leave it empty, for now.

Reviewer 2: At the same time you're missing 3 important section that are required:
- unit tests
- integration tests
- e2e tests
see the template: https://github.com/kubernetes/enhancements/blob/master/keps/NNNN-kep-template/README.md#test-plan

the general logic is:

when [performing update](https://github.com/kubernetes/kubernetes/blob/c984d53b31655924b87a57bfd4d8ff90aaeab9f8/pkg/controller/statefulset/stateful_set_control.go#L97-L138),
if the `currentRevision` != `updateRevision`, then we take it as starting to rolling update, but because rolling update can not
finish in one reconciling loop, then we may have to track it, we have two approaches:

- One is add a new field in the `defaultStatefulSetControl`, like below:

  ```golang
  type defaultStatefulSetControl struct {
    podControl        *StatefulPodControl
    statusUpdater     StatefulSetStatusUpdaterInterface
    controllerHistory history.Interface
    recorder          record.EventRecorder

    // <Newly Added>
    // rollingUpdateStartTimes records when each statefulset starts to perform rolling update.
    // key is the statefulset name, value is the start time.
    rollingUpdateStartTimes map[string]time.Time
  }
  ```

  Then when `currentRevision` != `updateRevision`, we'll check whether the statefulset name
  exists in the `rollingUpdateStartTimes`, if exists, skip, if not, write the

>   start time.

Reviewer 3: This doesn't work across restarts of kube-controller-manager.

Author: Yes, then maybe this is not a proper approach. 🤔

  If we found rolling update completed in [updateStatefulSetStatus](https://github.com/kubernetes/kubernetes/blob/c984d53b31655924b87a57bfd4d8ff90aaeab9f8/pkg/controller/statefulset/stateful_set_control.go#L682-L701),
  then we'll report the metrics here and clear the statefulset from the `rollingUpdateStartTimes`.

- Another approach is add a new field to `StatefulSetStatus` as below:

  ```golang
  type StatefulSetStatus struct {

    // rollingUpdateStartTime will record when the statefulSet starts to rolling update.
    rollingUpdateStartTime *time.Time
  }
  ```

  The logic general looks the same, we'll update the time together with status update, so there's no
  extra api calls.

>   Prefer Opt-2 as we already record the rolling update messages in the status.

Reviewer 3: @soltysh

>   Prefer Opt-2 as we already record the rolling update messages in the status.

Author: > Also - what if I change the spec again before the first rolling upgrade finished?

According to current implementation, two rolling updates will be combined as one. But we can add more logics to reset the rolling update duration.

#### Alpha

- Feature implemented behind a feature flag
- Initial e2e tests completed and enabled

#### Beta

- Gather feedback from developers and surveys
- Complete features A, B, C
- Additional tests are in Testgrid and linked in KEP

#### GA

- N examples of real-world usage
- N installs
- More rigorous forms of testing—e.g., downgrade tests and scalability tests
- Allowing time for feedback

**Note:** Generally we also wait at least two releases between beta and
GA/stable, because there's no opportunity for user feedback, or even bug reports,
in back-to-back releases.

**For non-optional features moving to GA, the graduation criteria must include
[conformance tests].**

[conformance tests]: https://git.k8s.io/community/contributors/devel/sig-architecture/conformance-tests.md

#### Deprecation

- Announce deprecation and support policy of the existing flag
- Two versions passed since introducing the functionality that deprecates the flag (to address version skew)
- Address feedback on usage/changed behavior, provided on GitHub issues
- Deprecate the flag
-->

- Alpha: Initial support for maxUnavailable in StatefulSets added. Disabled by default with default value of 1.

> - Beta:  Enabled by default with default value of 1 with upgrade downgrade tested at least manually.

Reviewer 2: See the above template (in comments) and fill them accordingly with alpha and beta requirements.

### Upgrade / Downgrade Strategy

<!--
If applicable, how will the component be upgraded and downgraded? Make sure
this is in the test plan.

Consider the following in developing an upgrade/downgrade strategy for this
enhancement:
- What changes (in invocations, configurations, API use, etc.) is an existing
  cluster required to make on upgrade, in order to maintain previous behavior?
- What changes (in invocations, configurations, API use, etc.) is an existing
  cluster required to make on upgrade, in order to make use of the enhancement?
-->

We will default to 1 for maxUnavailable field in StatefulSet for backward compatibility

Downgrades

When downgrading from a release with this feature, to a release without maxUnavailable, there are two cases
- If maxUnavailable is greater than 1, there are two more cases:-
  - If you're rolling back to a release that doesn't have this field - then there is even no way to discover it

>   - If you're just disabling the feature (either together with downgrade to a release that has a field or without downgrade),the field should remain set

Reviewer 2: This is wrong assumption, if we disable the feature there's a requirement that the field is cleared, that's also what the current implementation does: see https://github.com/kubernetes/kubernetes/blob/bb67eb5bd237334d6d0343aa382b351002653404/pkg/registry/apps/statefulset/strategy.go#L120-L137

Author: Thanks for this, I just copy-paste the original docs.

       (unless someone will explicitly delete it later), but controller should ignore its behavior (and there shouldn't be a way to set it if the feature gate
       is switched off).
 - If maxUnavailable is less than equal to 1 -- in this case user wont see any difference in behavior -->

### Version Skew Strategy

> No.

Reviewer 2: You're coordinating between kube-apiserver and kube-controller-manager the enablement of this feature. Please describe what happens when each of these components has the feature on and off accordingly. I'd expect a matrix with 4 entries for each of the components  having the feature on and off.

Author: Make sense

Author: Make sense

- maxUnavailable greater than 1 with partition and staged pods greater than maxUnavailable
- maxUnavailable greater than 1 with partition and maxUnavailable greater than replicas

Coverage:

- `pkg/apis/apps/v1`:               `2023-05-26` - `71.7%`
- `pkg/apis/apps/v1beta2`:          `2023-05-26` - `76.7%`
- `pkg/apis/apps/validation`:       `2023-05-26` - `92.3%`
- `pkg/controller/statefulset`:     `2023-05-26` - `85.7%`

> - `pkg/registry/apps/statefulset`:  `2023-05-26` - `63.9%`

Reviewer: Is there a coverage goal? I'm not sure what to make of this.

Author: No, just enumerate the core package in the form as suggested.

##### Integration tests

<!--
This question should be filled when targeting a release.
For Alpha, describe what tests will be added to ensure proper quality of the enhancement.

For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
https://storage.googleapis.com/k8s-triage/index.html
-->

> We have controller tests in `pkg/controller/statefulset` cover all cases. What's more,
> rolling update is hard to simulate in integration tests because we want to verify the pod
> status but actually they're faked. But we'll have e2e tests instead.

Reviewer: ```suggestion
Controller tests in `pkg/controller/statefulset` cover all cases. Since pod statuses
are faked in integration tests, simulating rolling update is difficult. We're opting
for e2e tests instead.
```

- Feature implemented behind a feature flag
- Initial e2e tests completed and enabled

#### Beta

- Gather feedback from developers and surveys
- Complete features A, B, C
- Additional tests are in Testgrid and linked in KEP

#### GA

- N examples of real-world usage
- N installs
- More rigorous forms of testing—e.g., downgrade tests and scalability tests
- Allowing time for feedback

**Note:** Generally we also wait at least two releases between beta and
GA/stable, because there's no opportunity for user feedback, or even bug reports,
in back-to-back releases.

**For non-optional features moving to GA, the graduation criteria must include
[conformance tests].**

[conformance tests]: https://git.k8s.io/community/contributors/devel/sig-architecture/conformance-tests.md

#### Deprecation

- Announce deprecation and support policy of the existing flag
- Two versions passed since introducing the functionality that deprecates the flag (to address version skew)
- Address feedback on usage/changed behavior, provided on GitHub issues
- Deprecate the flag
-->

#### Alpha

- Initial support for maxUnavailable in StatefulSets added. Disabled by default with default value of 1.

#### Beta

> - Enabled by default with default value of 1 with upgrade downgrade tested at least manually.

Reviewer 3: +1 [although for upgrade/downgrade manual are fine]

### Upgrade / Downgrade Strategy

<!--
If applicable, how will the component be upgraded and downgraded? Make sure
this is in the test plan.

Consider the following in developing an upgrade/downgrade strategy for this
enhancement:
- What changes (in invocations, configurations, API use, etc.) is an existing
  cluster required to make on upgrade, in order to maintain previous behavior?
- What changes (in invocations, configurations, API use, etc.) is an existing
  cluster required to make on upgrade, in order to make use of the enhancement?
-->

#### Upgrade

- What changes (in invocations, configurations, API use, etc.) is an existing
  cluster required to make on upgrade, in order to maintain previous behavior?

>   - Nothing special, it's backward compatibility

Reviewer: This is new behavior brought on by a new field, I'm very confused why backwards compatibility is even mentioned here.

- What changes (in invocations, configurations, API use, etc.) is an existing
  cluster required to make on upgrade, in order to make use of the enhancement?
  - This feature is enabled by default in beta so you can configure the maxUnavailable to meet your requirements

#### Downgrade

when downgrade from a release `maxUnavailable` enabled to a release it's disabled:

- If we downgrade to a release even doesn't have this field, then there's no way to discover it.

> - Else, if we create statefulset with maxUnavailable configured, it will be dropped to nil automatically.

Reviewer 3: That's not what we're doing.
The convention we're using is:
- if you try to set this field, it's silently dropped
- if the object with this field set already exists, we maintain it as it is - it's just controller that should ignore it

Reviewer: +1

###### What happens if we reenable the feature if it was previously rolled back?

> We will restore the desired behavior for StatefulSets for which the maxunavailable field wasn't deleted after

Reviewer 2: ```suggestion
We will restore the desired behavior for StatefulSets for which the maxUnavailable field wasn't deleted after
```

the feature gate was disabled.

###### Are there any tests for feature enablement/disablement?

> yes, there are unit tests which make sure the field is correctly dropped

Reviewer 2: We usually expect integration tests, see other proposals in sig-apps territory these days. For example: https://github.com/kubernetes/enhancements/pull/3967

on feature enable and disabled

### Rollout, Upgrade and Rollback Planning

###### How can a rollout or rollback fail? Can it impact already running workloads?

A rollout or rollback of this feature can fail if there is a bug which causes the kube-apiserver or

> the kube-controller-manager to start crashing when the feature flag is enabled.

Reviewer 2: That kind of case should be verified in integration testing. This question is more of what can happen assuming the feature is used (ie. field set) but we turn the feature off. What impact it may have on the workload.

> Yes, it can impact already running workloads.

Reviewer 2: How?

maxUnavailable pods with the same identity and never more than maxUnavailable being deleted.

###### What specific metrics should inform a rollback?

When feature enabled but rolling update in a unexpected phenomenon like the update pods at a time is not equal to the
`maxUnavailable` value (we may have some other factors impacting this, but this should apply for most of the time) or

> rolling update in a unexpected order.

Reviewer 2: I'm not sure I understand what you mean with this sentence? Can you make it more clear?

maxUnavailable pods with the same identity and never more than maxUnavailable being deleted.

###### What specific metrics should inform a rollback?

When feature enabled but rolling update in a unexpected phenomenon like the update pods at a time is not equal to the
`maxUnavailable` value or rolling update in a unexpected order.

Or we can refer to the `rolling-update-duration` metric for observation, if it didn't decrease when setting the `maxUnavailable`
great than 1 or the duration increased abnormally, then we should rollback.

> 

Reviewer 4: Are there other factors that may increase the rolling update duration? For example, if maxUnavailable is set to 2 but for some reason the rolling upgrade takes longer than when maxUnavailable was 1.

Maybe we should just focus on metrics such as "number of unavailable pods" and compare it with maxUnavailable to inform a rollback.

Author: > Are there other factors that may increase the rolling update duration? For example, if maxUnavailable is set to 2 but for some reason the rolling upgrade takes longer than when maxUnavailable was 1.

Exactly we do, let me add an explanation like `We may have some other factors impacting this, but this should apply for most of the time.

> Maybe we should just focus on metrics such as "number of unavailable pods" and compare it with maxUnavailable to inform a rollback.

If you mean adding a new metric, then the metric value is equal to `maxUnavailable` for most of the time, then it has not significant values.

kubectl get statefulsets -o yaml | grep maxUnavailable
```

Or refer to the new metric `rolling-update-duration`, it should exist.

###### How can someone using this feature know that it is working for their instance?

With feature enabled, set the `maxUnavailable` great than 1, and pay attention to the rolling update pods at a time,
it should equal to the `maxUnavailable`.
Or when setting the `maxUnavailable` great than 1, the `rolling-update-duration` should decrease.

###### What are the reasonable SLOs (Service Level Objectives) for the enhancement?

> I think it has little relevance with SLOs, but rolling update at a very low speed which impacts the running services.

Reviewer 2: What kind of values of `rolling-update-duration` metric you expect to be normal, for example?

Author: A statefulset with 2 replicas is incomparable with a statefulset with 20 replicas, it makes more sense as self-comparison, like `maxUnavailable=4` should roll update more quickly then `maxUnavailable=2`.

Author: Actually, I don't know how long is slow, we don't have any measurement about this, and the update duration impacts by a lot of factors, e.g. the node pressure, even the container hooks.

Reviewer 2: You don't have to be perfect with the proposed SLO, but I think we need to expose that information. We can reach out to sig-instrumentation for advice when implementing the metric itself and update this enhancement afterward. But don't say it has little relevance :wink:

Author: Updated.

###### What are the SLIs (Service Level Indicators) an operator can use to determine the health of the service?

> None.

Reviewer 2: What about `rolling-update-duration` mentioned above?

Author: It only works during rolling update, do you think we should count for that?

Reviewer 2: I'll repeat my doubt about the chosen metric, I can't find it, although it seems like one that should be exposed in during beta promotion.

Author: Added the metric, it located at the implementation details.

###### What happens if we reenable the feature if it was previously rolled back?

> We will restore the desired behavior for StatefulSets for which the maxUnavailable field wasn't deleted after

Reviewer 2: I think this one needs more clarification, I can't fully understand if this is talking about recovering maxUnavailable value? Because if it does, then that's not possible, when we disable the feature we set the field to nil, so you can't get it back .

Author: I think the true behavior is 
`If we disable the feature-gate and reenable it again, then the `maxUnavailable` will take effect again,
that's say we'll restart to respect the value of `maxUnavailable` when rolling update.`
Updated.

the feature gate was disabled.

###### Are there any tests for feature enablement/disablement?

<!--
The e2e framework does not currently support enabling or disabling feature
gates. However, unit tests in each component dealing with managing data, created
with and without the feature, are necessary. At the very least, think about
conversion tests if API types are being modified.

Additionally, for features that are introducing a new API field, unit tests that
are exercising the `switch` of feature gate itself (what happens if I disable a
feature gate after having objects written with the new field) are also critical.
You can take a look at one potential example of such test in:
https://github.com/kubernetes/kubernetes/pull/97058/files#diff-7826f7adbc1996a05ab52e3f5f02429e94b68ce6bce0dc534d1be636154fded3R246-R282
-->

> Yes, there are unit tests which make sure the field is correctly dropped

Author: Will, add to the Test plan.

on feature enable and disabled, see [strategy tests](https://github.com/kubernetes/kubernetes/blob/23698d3e9f4f3b9738ba5a6fcefd17894a00624f/pkg/registry/apps/statefulset/strategy_test.go#L391-L417).

### Rollout, Upgrade and Rollback Planning

###### How can a rollout or rollback fail? Can it impact already running workloads?

<!--
Try to be as paranoid as possible - e.g., what if some components will restart
mid-rollout?

Be sure to consider highly-available clusters, where, for example,
feature flags will be enabled on some API servers and not others during the
rollout. Similarly, consider large clusters and how enablement/disablement
will rollout across nodes.
-->

This could happen when we downgrade the kube-controller-manager from maxUnavailable-enabled release to a disabled one,

> but it will not impact running workloads, because `maxUnavailable` only works in rolling update.

Reviewer 3: +1

Also the other question that I have is what happens if there is an in-progress rollout (with maxUnavailable>1) and we disable the feature....

But if a statefulset has a plenty of replicas, when rolling update, it will take more time comparing to

> `maxUnavailable` enabled with a number greater than 1.

Reviewer 2: Again, I think the removed statement was more accurate wrt behavior description.

Yes, it can impact already running workloads.
This could happen when we downgrade the kube-controller-manager from maxUnavailable-enabled release to a disabled one,
but it will not impact running workloads, because `maxUnavailable` only works in rolling update.

If a rolling update is in progress for a StatefulSet, while this feature is being enabled in kube-apiserver
and kube-controller-manager, the StatefulSet controller can run into corner cases where it will take longer
for the controller to converge. This will only happen if after enabling the feature, the customer also sets
maxUnavailable to a number greater than 1,  but the invariants and the logic will ensure that there are never more than
maxUnavailable pods with the same identity and never more than maxUnavailable being deleted.
But if a statefulset has a plenty of replicas, when rolling update, it will take more time comparing to
`maxUnavailable` enabled with a number greater than 1.

###### What specific metrics should inform a rollback?

> TODO when we reach Beta

Reviewer: What metrics are you introducing here?

> When `maxUnavailable` enabled, and we're rolling update a statefulset, the number of pods brought down should

Author: Added, it's more comprehensive then。

The controller-manager will see very negligible and almost un-notoceable increase in cpu usage.

###### Can enabling / using this feature result in resource exhaustion of some node resources (PIDs, sockets, inodes, etc.)?

No.

### Troubleshooting

###### How does this feature react if the API server and/or etcd is unavailable?
The RollingUpdate will fail or will not be able to proceed if etcd or apiserver is unavailable and
hence this feature will also be not be able to be used.

###### What are other known failure modes?

<!--
For each of them, fill in the following information by copying the below template:
  - [Failure mode brief description]
    - Detection: How can it be detected via metrics? Stated another way:
      how can an operator troubleshoot without logging into a master or worker node?
    - Mitigations: What can be done to stop the bleeding, especially for already
      running user workloads?
    - Diagnostics: What are the useful log messages and their required logging
      levels that could help debug the issue?
      Not required until feature graduated to beta.
    - Testing: Are there any tests for failure mode? If not, describe why.
-->

> In a multi-master setup, when the cluster has skewed CCM, the behaviors may different.

Reviewer 2: ```suggestion
In a multi-master setup, when the cluster has skewed kube-controller-manager, the behaviors may differ.
```

- [Failure mode brief description]
  - Detection: the `rolling-update-duration` didn't decrease when setting the `maxUnavailable` great than 1 or increased abnormally.
  - Mitigations: Disable the feature.
  - Diagnostics: Set the logger level great than 4.
  - Testing: No testing, because the rolling update duration is hard to measure, it can be impact by a lot of things,
  like the master performance.

###### What steps should be taken if SLOs are not being met to determine the problem?

## Implementation History

- KEP Started on 1/1/2019
- Implementation PR and UT by 8/30

> - Bump to beta at 2023-05-11

Reviewer 2: The preferred format is: 
- date in YYYY-MM-DD format: descriptions
see other enhancements.

## Drawbacks

> Slow rolling update.

Reviewer 2: What do you mean with slow rolling update? Why is this a drawback of this feature? Isn't that the goal to allow faster rolling update with more pods down at any time?

Author: Misunderstood again, thanks.

###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?

Yes, you can disable the feature-gate manually once it's in Beta.

###### What happens if we reenable the feature if it was previously rolled back?

If we disable the feature-gate and reenable it again, then the `maxUnavailable` will take effect again,

> that's say we'll restart to respect the value of `maxUnavailable` when rolling update.

Reviewer 2: My bad, sorry :sweat_smile:

https://github.com/kubernetes/kubernetes/pull/97058/files#diff-7826f7adbc1996a05ab52e3f5f02429e94b68ce6bce0dc534d1be636154fded3R246-R282
-->

There are unit tests which make sure the field is correctly dropped
on feature enable and disabled, see [strategy tests](https://github.com/kubernetes/kubernetes/blob/23698d3e9f4f3b9738ba5a6fcefd17894a00624f/pkg/registry/apps/statefulset/strategy_test.go#L391-L417).

Feature enablement/disablement test will also be added when graduating to Beta as [TestStatefulSetStartOrdinalEnablement](https://github.com/kubernetes/kubernetes/blob/23698d3e9f4f3b9738ba5a6fcefd17894a00624f/pkg/registry/apps/statefulset/strategy_test.go#L473)

### Rollout, Upgrade and Rollback Planning

###### How can a rollout or rollback fail? Can it impact already running workloads?

<!--
Try to be as paranoid as possible - e.g., what if some components will restart
mid-rollout?

Be sure to consider highly-available clusters, where, for example,
feature flags will be enabled on some API servers and not others during the
rollout. Similarly, consider large clusters and how enablement/disablement
will rollout across nodes.
-->

> It depends, like

Reviewer 3: I'm not sure this really answers the question.

Please correct me if I'm wrong, but in general:
- this doesn't affect running workload on rollout [the only expection is if the feature was already enabled in the past and there are StatefulSets with the new field set and now they are going through a rolling upgrade when we enable the feature-gate-again]

On rollback may understanding is unclear - what happens if my StatsefulSet is going through a rolling upgrade now?

Author: If you mean what will happen if we disable the feature during a statefulset's rolling-update, then it will wait for all the pods ready and rolling-udpate a pod one time.

  by the statefulset controller.  But for newly created statefulset, their `maxUnavailable` field will
  be dropped if requested a feature-gate disabled api-server.

- With this feature enabled and a statefulset running into the rolling-update, then we disabled the feature-gate,
  then this statefulset will be impacted as it will fall into the default behavior, updating a pod at one time.

###### What specific metrics should inform a rollback?

<!--
What signals should users be paying attention to when the feature is young
that might indicate a serious problem?
-->

- As a normal user, with `maxUnavailable` enabled, and we're rolling update a statefulset, the number of pods brought down should
equal to the `maxUnavailable`, if not, we should rollback.

- As a administrator(as well for normal user), I can check the new metric `rolling-update-duration-seconds`, if we enabled the feature with
greater than 1 `maxUnavailable` set for statefulsets, but didn't see the duration fall down, then we may have some problems here.
This is not precise because rolling-update duration can be impacted by several factors, e.g. pulling down a big, new image, but

> can be some indicators refer to.

Reviewer 3: So can we come up with some better metric for it?

We're reporting ReadyReplicas/AvailableReplicas in Status - can we expose a metric if a difference between these and spec.Replicas is more than maxUnavailable as an example?

@soltysh - for thoughts

Reviewer 2: I like the proposed metric a lot more than the rolling time.

Author: Sorry, what messages can we get from `ReadyReplicas != AvailableReplicas`, in my understanding, this only relates to another filed `minReadySeconds`.

Author: `spec.Replicas - status.readyReplicas` should always be less than or equal to `spec.maxUnavailable`, or this is a bug.
The metric makes little sense I think ...

So for the rolling time, it values not only for the maxUnavailable, it can help to measure how long the rolling update takes, providing data for decisions whether to rolling-update in parallel if possible.  It can be independent of the KEP actually.

Reviewer 3: > spec.Replicas - status.readyReplicas should always be less than or equal to spec.maxUnavailable, or this is a bug.

Which is exactly what we want - we want the metric that shows us if something isn't ok.

FWIW - for me the "rolling-update-duration-seconds" makes very little sense because how should I think about it? How should I compare my 3-replica hello-world with 1000-replica gigantitc java app with 10m startup?
If we want to use for seeing if everything is fine, we need the metric to be somehow denormalizable.

Author: > How should I compare my 3-replica hello-world with 1000-replica gigantitc java app with 10m startup

Yes this also makes me doubtable about the metric, ok, then I'll drop this idea and then it will be more simple I think. Thanks.

Reviewer 5: @wojtek-t these information are for the most part exposed in kube-state-metrics today: https://github.com/kubernetes/kube-state-metrics/blob/main/docs/statefulset-metrics.md. We could add maxUnavailable with this effort. Would that be acceptable from a PRR standpoint or is it limited to native metrics?

Reviewer 5: Just for info, kube-state-metrics doesn't follow Kubernetes metrics stability framework just yet, so there are less guarantees towards its metrics than the ones we have in k/k.

But, although that's the case today, it would still be interesting to know if having API related metrics defined in kube-state-metrics would be acceptable from a KEP standpoint since in most cases they are unfit to the Kubernetes codebase.

Reviewer 3: Just to follow-up on it.
We didn't come up with a very concrete conclusion, but the hand-wavy one was:
- for the very simple metrics that effectively monitor a trivial logic on k8s side, we're fine with ksm metrics [like number of objects with a given field etc.]
- for things where the logic of the feature is non-trivial, we actually want a native metric

Now, with this feature we get somewhere on the fence. Let's say we add "maxUnavailable" as a metric to ksm.
IIUC, we can then check the value that I suggested above (replicas-read > maxUnavailable) for a given statefulset purely based on ksm.
Can we easily check how many statefulsets are in a bad state in a given time?

Now thinking further about it, maybe the metric that I would actually want to see is something like
`stateful_set_in_bad_state`  [we obviously need a better name] with a label `reason`.

Now, for this feature we can set `reason=exceededMaxUnavailable` and increase the value of that metric at any time when we process a statefulset and we observe that replicase-ready >maxUnavailable".

That actually seems (a) simple enough to implement (b) generic enough for future extension (c) useful enough as it shows if something goes wrong.

Thoughts?

###### What are the reasonable SLOs (Service Level Objectives) for the enhancement?

<!--
This is your opportunity to define what "normal" quality of service looks like
for a feature.

It's impossible to provide comprehensive guidance, but at the very
high level (needs more precise definitions) those may be things like:
  - per-day percentage of API calls finishing with 5XX errors <= 1%
  - 99% percentile over day of absolute value from (job creation time minus expected
    job creation time) for cron job <= 10%
  - 99.9% of /health requests per day finish with 200 code

These goals will help you determine what you need to measure (SLIs) in the next
question.
-->

Rolling update duration is impacted by a lot of factors, like the container hooks, image size, network, container logics,
so what we can roughly know here is for each individual statefulset. So generally, if we set the `maxUnavailable` to X, then
the rolling update duration should reduce X times. Keep in mind that, this is not precise.

###### What are the SLIs (Service Level Indicators) an operator can use to determine the health of the service?

<!--
Pick one more of these and delete the rest.
-->

- [x] Metrics

>   - Metric name: rolling-update-duration-seconds

Reviewer 2: `rolling_update_duration_seconds`

kubectl get statefulsets -o yaml | grep maxUnavailable
```

> Or refer to the new metric `rolling-update-duration`, it should exist.

Reviewer 3: +1

> Or refer to the new metric `rolling-update-duration`, it should exist.

Author: Added to the implementation details.

it should equal to the `maxUnavailable`.
Or when setting the `maxUnavailable` great than 1, the `rolling-update-duration` should decrease.

###### What are the reasonable SLOs (Service Level Objectives) for the enhancement?

<!--
This is your opportunity to define what "normal" quality of service looks like
for a feature.

It's impossible to provide comprehensive guidance, but at the very
high level (needs more precise definitions) those may be things like:
  - per-day percentage of API calls finishing with 5XX errors <= 1%
  - 99% percentile over day of absolute value from (job creation time minus expected
    job creation time) for cron job <= 10%
  - 99.9% of /health requests per day finish with 200 code

These goals will help you determine what you need to measure (SLIs) in the next
question.
-->

I think it has little relevance with SLOs, but rolling update at a very low speed which impacts the running services.

###### What are the SLIs (Service Level Indicators) an operator can use to determine the health of the service?

<!--
Pick one more of these and delete the rest.
-->

- [x] Metrics
  - Component exposing the metric: kube-controller-manager

>     - Metric name: `rolling-update-duration{plugin="PodTopologySpread"}`

Reviewer: conventionally: `rolling_update_duration_seconds`

Author: thanks, updated.

hence this feature will also be not be able to be used.

###### What are other known failure modes?

<!--
For each of them, fill in the following information by copying the below template:
  - [Failure mode brief description]
    - Detection: How can it be detected via metrics? Stated another way:
      how can an operator troubleshoot without logging into a master or worker node?
    - Mitigations: What can be done to stop the bleeding, especially for already
      running user workloads?
    - Diagnostics: What are the useful log messages and their required logging
      levels that could help debug the issue?
      Not required until feature graduated to beta.
    - Testing: Are there any tests for failure mode? If not, describe why.
-->

Disable the feature-gate when performing rolling-update.

- [Failure mode brief description]

>   - Detection: The `maxUnavailabel` is not respected.

Reviewer 2: ```suggestion
  - Detection: The `maxUnavailable` is not respected.
```
